File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp

class Solution {
public:
    // int maxlen = 0;
    int dfs(TreeNode* root, int& maxlen) {
        if (!root)
            return -1;
        int left_len = dfs(root->left, maxlen) + 1;
        int right_len = dfs(root->right, maxlen) + 1;
        maxlen = max(maxlen, left_len + right_len);
        return max(left_len, right_len);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int maxlen = 0;
        dfs(root, maxlen);
        return maxlen;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
maxDiameter = 0

class Solution:
    def getDepth(self, root: Optional[TreeNode]) -> int:
        global maxDiameter
        if not root.left and not root.right:
            return 0
        leftSubtreeDepth = self.getDepth(root.left) + 1 if root.left else 0
        rightSubtreeDepth = self.getDepth(root.right) + 1 if root.right else 0
        maxDiameter = max(maxDiameter, leftSubtreeDepth + rightSubtreeDepth)
        return max(leftSubtreeDepth, rightSubtreeDepth)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        global maxDiameter
        maxDiameter = 0
        return max(self.getDepth(root), maxDiameter)